Item 9758: Field editor "Text Choice" data type and UI to create, update, delete field options/values

cnathe opened the pull request:
#### Rationale
The previous story in this epic added a TextChoiceValidator to the server side and updated the UI for the field insert/update forms to show a select input of the values. This PR is part 1 of updating the field editor UI to add the new "Text Choice" data type and the UI in the field editor expanded row to define the set of values / choices for that field. Part 2 of the Text Choice UI will handle the "changing of a choice value that is already in use", but for this first phase it just sets those options as locked when they are in use.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/692
* https://github.com/LabKey/platform/pull/2876
* https://github.com/LabKey/sampleManagement/pull/784
* https://github.com/LabKey/biologics/pull/1098
* https://github.com/LabKey/commonAssays/pull/400

#### Changes
* factor out ChoicesListItem to re-use in list group cases (TextChoiceOptions, ChoosePicklistModal, ManageSasmpleStatusesPanel)
* field editor row addition of "Text Choice" data type and "Text Choice Options" section in expanded row
* field editor row updates for selected "Text Choice" data type: reset some field options set for other types, hide regex validator UI, hide text options UI
* factor out DomainRowWarning to re-use between name special character type warnings and empty text choice values details display
* TextChoiceAddValuesModal for add case and check for max allowed values
* domain design model parsing for loading existing field property validators for text choice fields

cnathe committed:
RegEx validator modal fix for “Fail validation when pattern matches field value” checked state on initialization

cnathe committed:
Add new Text Choice data type and concept URI

cnathe committed:
v2.103.0-fb-textChoiceUI.0

cnathe committed:
Merge branch 'develop' into fb_textChoiceUI

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.103.1-fb-textChoiceUI.0

cnathe committed:
jest snapshot update

cnathe committed:
Factor out ChoicesListItem to use with add to picklist modal and manage samlpe statuses

cnathe committed:
TextChoiceOptions component CRUD
- display of values in list group
- allow adding via modal text area input, up to max of 200
- allow removal of values
- allow edit of values

cnathe committed:
Domain field behavior for change to and from Text Choice data type
- add default textChoiceValidator on change to and remove on change from

cnathe committed:
v2.103.1-fb-textChoiceUI.1

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
add schemaName and queryName to DomainDesign and pass through values to TextChoiceOptions.tsx component

cnathe committed:
TextChoiceOptions query to get distinct values for existing column
- to be used as initial values when changing a Text field to Text Choice
- to be used for marking Text Choice values as locked (i.e. in use)

cnathe committed:
factor out TextChoiceOptionsImpl component now, which will make jest testing easier

cnathe committed:
v2.103.1-fb-textChoiceUI.2

cnathe committed:
Text choice field editor updates for details display and warning
- show field warning if no choice values exist for the field
- spit out DomainRowWaring component
- show field details for text choice values, or warning if none present

cnathe committed:
v2.103.1-fb-textChoiceUI.3

cnathe committed:
Jest snapshot updates

cnathe committed:
Jest test fix for getDetailsTextArray() addition of index param

cnathe committed:
Jest test fix for DomainForm data type options length

cnathe committed:
Jest snapshot fix

cnathe committed:
Only show text choice field warning for empty values after initial values loaded / added

cnathe committed:
Use original field name when querying for distinct values on change to Text Choice field data type

cnathe committed:
Trim text choice value when checking for valid and on apply

cnathe committed:
v2.103.1-fb-textChoiceUI.4

cnathe committed:
Even more jest snapshot updates

cnathe committed:
When field is changed to Text Choice clear other validators that might have been added to the field

cnathe committed:
Don't show Text Choice values in query metadata editor view, which does not show validator info

cnathe committed:
Fixes for importing fields from JSON to field editor
- clear validator rowId on JSON import so that they save correctly
- set Text Choice values based on expression on JSON import

cnathe committed:
v2.103.1-fb-textChoiceUI.5

cnathe committed:
reset text field scale value on change to Text Choice

cnathe committed:
jest snapshot fix

cnathe committed:
v2.103.1-fb-textChoiceUI.6

cnathe committed:
update release notes

cnathe committed:
update text choice value right hand panel display / layout for large screen

cnathe committed:
Merge branch 'develop' into fb_textChoiceUI

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md
#	packages/components/src/internal/components/domainproperties/DomainRowExpandedOptions.tsx
#	packages/components/src/internal/components/domainproperties/actions.ts

cnathe committed:
v2.105.0-fb-textChoiceUI.0

labkey-susanh commented on packages/components/src/internal/components/domainproperties/PropDescType.ts:
```diff
@@ -75,6 +76,10 @@ export class PropDescType
     }
 
+    static isTextChoice(name: string): boolean {
+        return name === TEXT_CHOICE_CONCEPT_URI || name === 'textChoice';
```
When will we have `name === 'textChoice'` instead of the concept URI?

cnathe replied:
For the field row data type select input on change event, looks like it passes the field input name prop instead of the concept URI here (so that it can apply for non-conceptURI based data types).

labkey-susanh replied:
Is it worth adding a comment about that?

cnathe replied:
Done

labkey-susanh commented on packages/components/src/internal/components/domainproperties/TextChoiceAddValuesModal.tsx:
```diff
@@ -0,0 +1,74 @@
+import { MAX_VALID_TEXT_CHOICES } from './constants';
+
+interface Props {
+    title: string;
```
I wonder if this would be better called `fieldName` since it is only part of the title.

cnathe replied:
Good point. Done.

labkey-susanh commented on packages/components/src/internal/components/domainproperties/TextChoiceAddValuesModal.tsx:
```diff
@@ -0,0 +1,74 @@
+                <p>
+                    {`A total of ${Utils.pluralize(maxValuesToAdd, 'value', 'values')} can be added. `}
+                    <span className={classNames({ 'domain-text-choices-error': parsedValues.length > maxValuesToAdd })}>
+                        The current text input includes {Utils.pluralize(parsedValues.length, 'value', 'values')}.
```
Perhaps "There are currently X new text choice values provided below." or "Currently, X new text choice values have been provided below."  The reference to "text input" threw me a bit and seems kinda techy.

cnathe replied:
Done

labkey-susanh commented on packages/components/src/internal/components/domainproperties/TextChoiceOptions.tsx:
```diff
@@ -0,0 +1,282 @@
+    const onSelect = useCallback(
+        ind => {
+            setSelectedIndex(ind);
+            setCurrentValue(validValues[ind]);
```
Any chance that `validValues` is undefined?

cnathe replied:
I don't believe so, but doesn't hurt to check. Done.

labkey-susanh commented on packages/components/src/internal/components/domainproperties/TextChoiceOptions.tsx:
```diff
@@ -0,0 +1,282 @@
+                                        index={ind}
+                                        key={ind}
+                                        label={value}
+                                        itemType={value === '' && 'Empty Value'}
```
The use of 'Empty Value' or `false` as `itemType` confuses me.  These don't seem compatible as "types" to me.

cnathe replied:
yeah, `itemType` was a hold over from the property name when I factored this component out of the manage sample type statuses component use case. I've changed it to `subLabel`.

labkey-susanh commented on packages/components/src/internal/components/domainproperties/TextChoiceOptions.tsx:
```diff
@@ -0,0 +1,282 @@
+            onChange(
+                fieldId,
+                new PropertyValidator({
+                    // keep the existing validator Id/props, if present, and override teh expression / properties
```
```suggestion
                    // keep the existing validator Id/props, if present, and override the expression / properties
```

cnathe replied:
done

labkey-susanh commented on packages/components/src/internal/components/domainproperties/TextChoiceOptions.tsx:
```diff
@@ -0,0 +1,282 @@
+
+                        // if this is an existing text field that is being changed to a Text Choice data type,
+                        // then we will use the existing distinct values for that field as the initial options
+                        if (!field.textChoiceValidator?.rowId) {
```
The comment is explaining the opposite of what the if is testing, so seems confusing to me.

cnathe replied:
done, updated to hopefully be less confusing.

labkey-susanh commented on packages/components/src/internal/components/domainproperties/dataclasses/__snapshots__/DataClassDesigner.spec.tsx.snap:
```diff
@@ -1268,6 +1269,7 @@ exports[`DataClassDesigner initModel 1`] = `
                 "lookupSchema": undefined,
                 "lookupQuery": undefined,
               },
+              "name": "TextFIeld",
```
Is the casing here intentional?

cnathe replied:
argh, nope not intentional. Fixed.

labkey-susanh commented on packages/components/src/internal/components/domainproperties/models.tsx:
```diff
@@ -537,24 +547,41 @@ export class ConditionalFormat
+
+    constructor(values?: { [key: string]: any }) {
+        if (typeof values?.failOnMatch === 'string') {
+            values.failOnMatch = values.failOnMatch === 'true';
```
Should this be case-insensitive?

cnathe replied:
Done

labkey-susanh commented on packages/components/src/internal/components/domainproperties/validation/RegexValidationOptions.tsx:
```diff
@@ -216,12 +136,64 @@ export class RegexValidationOptions extends React.PureComponent<RegexValidationO
+                                        regular expression. Check this box if you want validation to fail when the
+                                        pattern matches the field value.
+                                    </LabelHelpTip>
+                                </Checkbox>
```
Are we offering this option for the TextChoice fields as well?  Seems odd.

cnathe replied:
Nope, we aren't offering this in the UI in any way. If you view the JSON for a text choice validator it will have this property since it is included in the default validator definition, but it has no impact on the text choice case.

labkey-susanh reviewed (approved):
Looks good.  Just some minor comments and suggestions here.  No need for re-review

cnathe committed:
Domain allowTextChoiceProperties property

cnathe committed:
jest snapshot updates

cnathe committed:
Jest test update for allowTextChoiceProperties

cnathe committed:
v2.105.0-fb-textChoiceUI.1

cnathe committed:
Misc code review feedback

cnathe committed:
v2.105.0-fb-textChoiceUI.2

cnathe committed:
Merge branch 'develop' into fb_textChoiceUI

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.106.0-fb-textChoiceUI.0

cnathe committed:
Show TextChoiceInput in editable grid cell for column with validValues

cnathe committed:
update release notes

cnathe committed:
Jest test fix for placeholder default value change

cnathe committed:
v2.106.0-fb-textChoiceUI.1

cnathe committed:
Jest tests for updates related to Text Choice data type UI

cnathe committed:
Jest test for LookupCell

cnathe committed:
Jest test for DomainRowWarning

cnathe committed:
Jest test for TextChoiceAddValuesModal

cnathe committed:
Jest test for TextChoiceOptions

cnathe committed:
v2.106.0-fb-textChoiceUI.2

cnathe committed:
Modal textarea full width

cnathe committed:
v2.106.0-fb-textChoiceUI.3

cnathe committed:
Misc fixes from manual testing notes
- update text choice related help tip text
- add DisableableButton component and use in various places
- add css rule for textarea-fullwidth
- editable grid lookup cell hover styling for chevron icon
- remove empty strings from add values modal parsing

cnathe committed:
Fix jest tests for TextChoiceAddValuesModal.tsx changes

cnathe committed:
v2.106.0-fb-textChoiceUI.4

cnathe committed:
Wrap text in text choice list items and field editor row details

cnathe committed:
Text choice row details display fixes
- remove duplciates from details text and count
- at > 80 characters, just show the total count of values

cnathe committed:
v2.106.0-fb-textChoiceUI.5

cnathe committed:
Text Choice add values text updates to remove redundant text

cnathe committed:
Text Choice add values modal update to remove duplicates from parsed value array

cnathe committed:
v2.106.0-fb-textChoiceUI.6

cnathe committed:
jest updates / additions for recent changes / fixes

cnathe committed:
Merge branch 'develop' into fb_textChoiceUI

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
update release notes date and version

cnathe committed:
v2.108.0
